x/rollapp/genesis.go

```diff
@@ -40,10 +66,89 @@ func ExportGenesis(ctx sdk.Context, k keeper.Keeper) *types.GenesisState {
 
 	genesis.RollappList = k.GetAllRollapps(ctx)
 	genesis.StateInfoList = k.GetAllStateInfo(ctx)
-	genesis.LatestStateInfoIndexList = k.GetAllLatestStateInfoIndex(ctx)
-	genesis.LatestFinalizedStateIndexList = k.GetAllLatestFinalizedStateIndex(ctx)
-	genesis.BlockHeightToFinalizationQueueList = k.GetAllBlockHeightToFinalizationQueue(ctx)
 	// this line is used by starport scaffolding # genesis/module/export
 
 	return genesis
 }
+
+func removeAllGenesisState(ctx sdk.Context, k keeper.Keeper) {
+	for _, rollapp := range k.GetAllRollapps(ctx) {
+		k.RemoveRollapp(ctx, rollapp.RollappId)
+	}
+	for _, stateInfo := range k.GetAllStateInfo(ctx) {
+		k.RemoveStateInfo(ctx, stateInfo.StateInfoIndex.RollappId, stateInfo.StateInfoIndex.Index)
+	}
+}
+
+func checkAllRollapsStateInfo(ctx sdk.Context, k keeper.Keeper) error {
+	rollappsList := k.GetAllRollapps(ctx)
+
+	for _, rollapp := range rollappsList {
+		stateInfoList := k.GetAllRollappStateInfo(ctx, rollapp.RollappId)
+		previousIndex := uint64(0)
+		lastFinalized := uint64(0)
+		prevStateInfoLastBlock := uint64(0)
+		stateInfoIsCorrect := true
+		var errormsg string
+		for _, stateInfo := range stateInfoList {
+			if stateInfo.StateInfoIndex.Index != previousIndex+1 {
+				errormsg = "missing state info"
+				stateInfoIsCorrect = false
+				break
+			}
+			if stateInfo.Status == common.Status_FINALIZED {
+				if lastFinalized != stateInfo.StateInfoIndex.Index-1 {
+					errormsg = "inconsistent finalized status"
+					stateInfoIsCorrect = false
+					break
+				}
+				lastFinalized = stateInfo.StateInfoIndex.Index
+			}
+
+			if stateInfo.StartHeight-1 != prevStateInfoLastBlock {
+				errormsg = "missing block in state info"
+				stateInfoIsCorrect = false
+				break
+			}
+			prevStateInfoLastBlock = stateInfo.GetLatestHeight()
+			previousIndex = stateInfo.StateInfoIndex.Index
+		}
+		if !stateInfoIsCorrect {
+			removeAllGenesisState(ctx, k)
+			return errorsmod.Wrapf(sdkerrors.ErrLogic, "error init genesis validating state info rollapp:%s: error:%s", rollapp.RollappId, errormsg)
+		}
+		if previousIndex != uint64(0) {
+			k.SetLatestStateInfoIndex(ctx, types.StateInfoIndex{RollappId: rollapp.RollappId, Index: previousIndex})
+		}
+		if lastFinalized != uint64(0) {
+			k.SetLatestFinalizedStateIndex(ctx, types.StateInfoIndex{RollappId: rollapp.RollappId, Index: lastFinalized})
+		}
+	}
+	return nil
+}
+
+func buildBlockHeightToFinalizationQueue(ctx sdk.Context, k keeper.Keeper) {
+	blockHeightToFinalizationQueue := make(map[uint64][]types.StateInfoIndex)
+	rollappsList := k.GetAllRollapps(ctx)
+
+	for _, rollapp := range rollappsList {
+		stateInfoList := k.GetAllRollappStateInfo(ctx, rollapp.RollappId)
+
+		for _, stateInfo := range stateInfoList {
+			if stateInfo.Status == common.Status_PENDING {
+				// load FinalizationQueue and update
+				finalizationQueue, found := blockHeightToFinalizationQueue[stateInfo.CreationHeight]
```

Comment:
don't think this part makes sense logically. the reason is that if we import genesis we can assume starting a chain from scratch. in that sense the height of the chain will start from 1. 

Now if you import a pending state from a previous rollapp, you can't rely on it's previous creation height as it's relative to the other chain it was imported from.

Assume the other chain had height 1M for the state info, it menas this state info will get finalized only in 1M blocks.